docs/proposals/compute-resource-metrics-api.md

```diff
@@ -0,0 +1,179 @@
+
+* Push - each node would periodically POST their derived and raw metrics to
+  the API server. Kubectl and the schedule would watch the API server to get
+  the latest metrics.
+
+* Poll - heapster would periodically poll a kubelet API endpoint on each node
+  to fetch its node and pods metrics. Then Heapster will have to implement the
+  APIs mentioned above.
+
+Our initial plan is to follow the push based system.
+
+The API server will not hold any of this data directly. For our initial
+purposes, it will hold the most recent metrics obtained from each node in
+etcd. Then, when polled for metrics, the API server would only serve said most
+recent data per node. The clients could cache this data if they needed to keep
+a historical record of it, for example Heapster.
+
+In the future, we might want to store the metrics differently:
+
+* via heapster - Since heapster keeps data for a period of time, we could
+  redirect requests to the API server to heapster instead of using etcd. This
+  would also allow serving metrics other than the latest ones.
+
+## Data structure
+
+```Go
+type DerivedPodMetrics struct {
+	TypeMeta
+	ObjectMeta // should have pod name
+	// the key is the container name
+	Containers map[string]ResouresMetrics
+}
+
+type DerivedNodeMetrics struct {
+	TypeMeta
+	ObjectMeta // should have node name
+	NodeMetrics      ResourcesMetrics
+	SystemContainers map[string]ResourcesMetrics
+}
+
+type ResourcesMetrics map[resource.Type]DerivedMetrics
+
+type DerivedMetrics struct {
+	// Time at which these metrics were gathered
+	Time util.Time `json:"timestamp"`
+
+	// 1m, 1h and 1d as a start
+	Metrics map[time.Duration]Metrics
+}
+
+type Metrics struct {
+	// Average over the collected sample.
+	Mean resource.Quantity `json:"mean"`
+	// Max seen over the collected sample.
+	Max resource.Quantity `json:"max"`
+	// 95th percentile over the collected sample.
+	NinetyFive resource.Quantity `json:"95th"`
+}
+
+type RawPodMetrics struct {
+	TypeMeta
+	ObjectMeta // should have pod name
+	// the key is the container name
+	Containers map[string]struct {
+		ContainerReference *Container
+		RawMetrics         []RawMetric
+	}
+}
+
+type RawNodeMetrics struct {
+	TypeMeta
+	ObjectMeta // should have node name
+	NodeMetrics      []RawMetric
+	SystemContainers map[string][]RawMetric
+}
+
+type RawMetric {
+	Time util.Time
+	CPU  CpuStats
+	Mem  MemoryStats
+}
+
+// All CPU usage metrics are cumulative from the creation of the container
+type CpuStats struct {
+	Usage CpuUsage `json:"usage"`
+	// Smoothed average of number of runnable threads x 1000.
+	// We multiply by thousand to avoid using floats, but preserving precision.
+	// Load is smoothed over the last 10 seconds. Instantaneous value can be read
+	// from LoadStats.NrRunning.
+	LoadAverage int32 `json:"load_average"`
+}
+
+// CPU usage time statistics.
+type CpuUsage struct {
+	// Total CPU usage.
+	// Units: cores
+	Total resource.Quantity `json:"total"`
```

Comment:
On Jul 28, 2015, at 5:30 PM, Vish Kannan <notifications@github.com> wrote:

In docs/proposals/compute-resource-metrics-api.md
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11951#discussion_r35703738>
:

> +
> +// All CPU usage metrics are cumulative from the creation of the container
> +type CpuStats struct {
> +	Usage CpuUsage `json:"usage"`
> +	// Smoothed average of number of runnable threads x 1000.
> +	// We multiply by thousand to avoid using floats, but preserving precision.
> +	// Load is smoothed over the last 10 seconds. Instantaneous value can be read
> +	// from LoadStats.NrRunning.
> +	LoadAverage int32 `json:"load_average"`
> +}
> +
> +// CPU usage time statistics.
> +type CpuUsage struct {
> +	// Total CPU usage.
> +	// Units: cores
> +	Total resource.Quantity `json:"total"`

Some of these metrics are made available by the kernel. These are in
addition to the basic resources that the user requests like cpu, memory,
filesystem, etc. There is value for these additional metrics for
functionalities like monitoring, auto-pilot, etc. But these 'raw' metrics
don't correspond one-to-one with the resource types specified as part of a
pod or container. We can either expose these metrics as part of the core
API or choose to have a separate metrics control plane (kubelet, heapster,
etc).


Is it reasonable to expect that every Kube admin would expect to have
access to these metrics?

Part of the reason why I ask about the resource schema is whether I would
deal with these metrics generically (like we do with requests and limits
for quota).

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11951/files#r35703738>
.